client/dwh_migration_client/macro_expander.py

```diff
@@ -0,0 +1,273 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Macro handling classes """
+
+import logging
+import re
+from fnmatch import fnmatch
+from typing import Callable, Dict, List, Optional, Set, Tuple
+
+
+class RecordingLogger:
+    def __init__(self) -> None:
+        self.messages = []
+
+    """Logger that keeps track of the messages that it has logged."""
+
+    def warnLog(self, format_string, *args) -> None:
+        message = format_string.format(*args)
+        self.messages.append(message)
+        logging.warning(message)
+
+
+class MacroExpander(RecordingLogger):
+    """Base class for macro expanders. Do not use this directly."""
+
+    # macro name -> replacement
+    mapping: Optional[Dict[str, str]]
```

Comment:
I see you are declaring uninitialized class vars with type annotations for this class and others. Is there a benefit to doing this? I only do this for data classes decorated with `@dataclasses.dataclass` as the decorator will essentially erase the class vars and generate an `__init__` constructor with equivalent instance vars. For a normal class, you should be able to just assign instance variables manually in `__init__` with type annotations only being required when they can't be inferred by `mypy`.

Reply:
Fixed.